app/controllers/admin/users_controller.rb

```diff
@@ -69,6 +78,18 @@ def destroy
       redirect_to admin_user_path(params[:id])
     end
 
+    def export
+      @users = User.select(ATTIBUTES_FOR_CSV + ATTRIBUTES_FOR_LAST_ACTIVITY).includes(:organizations)
```

Comment:
Looks like `User#has_role?` invokes N+1 queries on the `user_status` check, and adding `includes(:roles)` doesn't fix it. It looks like that method [necessarily fires off a query](https://github.com/RolifyCommunity/rolify/blob/0c883f4173f409766338b9c6dfc64b0fc8ec8a52/lib/rolify/role.rb#L36) and does not check whether the `user.roles` association is already loaded. In order to make use of eager loading, we need to use [`has_cached_role?`](https://github.com/RolifyCommunity/rolify/blob/0c883f4173f409766338b9c6dfc64b0fc8ec8a52/lib/rolify/role.rb#L47-L50).

The snippet below (note: `has_cached_role?` in `user_status`, the `includes` call now loads `roles`, and I switched from `each` to `find_each`) ran in 3:11.727 in a Rails console on DEV, generating a 78MB CSV file. This is still about 6x too long for DEV but it's good to have at least a benchmark.

<details>
<summary>Click to see code</summary>

```ruby
def user_status(user)
  if user.has_cached_role?(:suspended)
    "Suspended"
  elsif user.has_cached_role?(:warned)
    "Warned"
  elsif user.has_cached_role?(:comment_suspended)
    "Comment Suspended"
  elsif user.has_cached_role?(:trusted)
    "Trusted"
  else
    "Good Standing"
  end
end

columns = %i[
  id
  name
  username
  email
  registered_at
  registered
  last_comment_at
  last_article_at
  latest_article_updated_at
  last_reacted_at
  profile_updated_at
  last_moderation_notification
  last_notification_activity
]
Benchmark.realtime do
  File.open("tmp/users.csv", "w") do |f|
    csv = CSV.new(f)
    User
      .select(columns)
      .includes(:organizations, :roles)
      .find_each do |user|
        csv << [
          user.name,
          user.username,
          user.email,
          user_status(user),
          user.registered_at&.strftime("%d %b, %Y"),
          user.last_activity&.strftime("%d %b, %Y"),
          user.organizations.pluck(:name),
        ]
      end
  end
end
```

</details>

We might be able to trim this down a bit if we got a bit craftier with the SQL query. That would take some experimentation, though, since it could actually end up being slower. We'd have to use `LEFT JOIN`s (not everyone would have roles or organizations), which [can be slow](https://dev.to/jgaskins/tricksy-little-postgres-query-planner-42hg). Regardless, I don't think we'd be able to bring it down below 30 seconds, so I'm not sure if spending that much time is useful here. The biggest performance win seems top be figuring out how to incorporate `has_cached_role?` so we don't issue on the order of 1M `user.has_role?` queries here. The instrumentation alone on that many queries would probably run the Sidekiq worker out of memory. 😄

Reply:
This is a good catch, looking forward to working on the performance part with you!